src/core/shared_memory_manager.cc

```
@@ -0,0 +1,244 @@
+//
+
+#include "src/core/shared_memory_manager.h"
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <algorithm>
+#include <deque>
+#include <exception>
+#include <future>
+#include "src/core/constants.h"
+#include "src/core/logging.h"
+#include "src/core/server_status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+Status
+open_shm_region(const std::string& shm_key, int* shm_fd)
+{
+  // get shared memory region descriptor
+  *shm_fd = shm_open(shm_key.c_str(), O_RDONLY, S_IRUSR | S_IWUSR);
+  if (*shm_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to get shared memory descriptor");
+  }
+
+  return Status::Success;
+}
+
+Status
+get_shm_addr(
+    const int shm_fd, const size_t offset, const size_t byte_size,
+    void** shm_addr)
+{
+  // map shared memory to process address space
+  *shm_addr = mmap(NULL, byte_size, PROT_WRITE, MAP_SHARED, shm_fd, offset);
+  if (*shm_addr == MAP_FAILED) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to process address space");
+  }
+
+  return Status::Success;
+}
+
+Status
+shm_close(int shm_fd)
+{
+  int tmp = close(shm_fd);
+  if (tmp == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to close shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+mmap_cleanup(void* shm_addr, size_t byte_size)
+{
+  int tmp_fd = munmap(shm_addr, byte_size);
+  if (tmp_fd == -1) {
+    return Status(
+        RequestStatusCode::INTERNAL, "Unable to munmap shared memory region");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::RegisterSharedMemory(
+    const std::string& name, const std::string& shm_key, const size_t offset,
+    const size_t byte_size)
+{
+  LOG_VERBOSE(1) << "Register() shared memory region: '" << name << "'";
+  Status status = Status::Success;
+
+  // check if key is in shared_memory_map_ then remove
+  if (shared_memory_map_.find(name) != shared_memory_map_.end()) {
```

This can be dangerous if client makes an mistake are reuses name for 2 different regions. Instead we should return an error.